google-cloud-firestore/src/test/java/com/google/cloud/firestore/RecursiveDeleteTest.java

```diff
@@ -0,0 +1,733 @@
+                FieldPath.documentId().toString(),
+                recursiveDeleteStartAt("coll")),
+            filter(
+                Operator.LESS_THAN,
+                FieldPath.documentId().toString(),
+                recursiveDeleteEndAt("coll")),
+            limit(RecursiveDelete.MAX_PENDING_OPS));
+
+    firestoreMock.recursiveDelete(firestoreMock.collection("coll")).get();
+    assertEquals(expectedRequest, runQueryCapture.getAllValues().get(1));
+  }
+
+  @Test
+  public void createsSecondQueryWithCorrectStartAfter() throws Exception {
+    // This test checks that the second query is created with the correct startAfter() once the
+    // RecursiveDelete instance is below the MIN_PENDING_OPS threshold to send the next batch.
+
+    // Use lower limits than the actual RecursiveDelete class in order to make this test run fast.
+    int maxPendingOps = 100;
+    int minPendingOps = 11;
+    int maxBatchSize = 10;
+
+    final int cutoff = maxPendingOps - minPendingOps;
+    final int[] numDeletesBuffered = {0};
+
+    // This future is used to delay the BatchWriteResponses from returning in order to create the
+    // situation where the number of pending operations is less than `minPendingOps`.
+    final SettableApiFuture<Void> bufferFuture = SettableApiFuture.create();
+
+    // This future completes when the second query is run.
+    final SettableApiFuture<Void> secondQueryFuture = SettableApiFuture.create();
+
+    List<String> firstStream = new ArrayList<>();
+    final List<ApiFuture<BatchWriteResponse>> batchWriteResponse = new ArrayList<>();
+    for (int i = 0; i < maxPendingOps; i++) {
+      firstStream.add(fullDocumentPath("doc" + i));
+    }
+
+    for (int i = 0; i < maxBatchSize; i++) {
+      batchWriteResponse.add(successResponse(1));
+    }
+
+    doAnswer(queryResponse(firstStream.toArray(new String[0])))
+        .doAnswer(
+            new Answer<RunQueryResponse>() {
+              public RunQueryResponse answer(InvocationOnMock invocation) throws Throwable {
+                secondQueryFuture.set(null);
+                Object[] args = invocation.getArguments();
+                ApiStreamObserver<RunQueryResponse> observer =
+                    (ApiStreamObserver<RunQueryResponse>) args[1];
+                observer.onCompleted();
+                return null;
+              }
+            })
+        .when(firestoreMock)
+        .streamRequest(
+            runQueryCapture.capture(),
+            streamObserverCapture.capture(),
+            Matchers.<ServerStreamingCallable>any());
+
+    doAnswer(
+            new Answer<ApiFuture<BatchWriteResponse>>() {
+              public ApiFuture<BatchWriteResponse> answer(InvocationOnMock mock) throws Exception {
+                if (numDeletesBuffered[0] < cutoff) {
+                  numDeletesBuffered[0] += batchWriteResponse.size();
+                  return ApiFutures.transformAsync(
+                      bufferFuture,
```

Comment:
I don't quite understand why we need to block on `bufferFuture` here. I thought the idea was that we should finish the first 4000 deletes and then ask for the next batch. The current test seems to indicate that we do not wait for the deletes to finish executing.

Reply:
Thanks for explaining this. I am pretty sure I understand this now. If you can add a small comment in the test I might even still understand tomorrow (optional).

Reply:
Added a comment.